File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
    
    const row = maps.length
    const col = maps[0].length
    const visited= Array.from({length : row}, () => Array(col).fill(false))
    
    const isInGrid = (nr,nc) => {
        return (nr>=0 && nr<row && nc>=0 && nc<col )
    }
    

    
    const bfs  = (maps)=>{
        const queue = [[0,0,1]]
        const dr = [-1, 1, 0, 0] // t b l r
        const dc = [0, 0, -1, 1]
        
        while (queue.length > 0){
            const [r, c, dist] = queue.shift()
            if (r === row -1 && c === col -1 ){
                return dist
            }
            for(let i=0; i<4; i++){
                const nr = r+ dr[i]
                const nc = c+ dc[i]
                
                if(isInGrid(nr, nc) && maps[nr][nc] != 0 && !visited[nr][nc]){
                    visited[nr][nc] = true
                    queue.push([nr,nc,dist+1])
                }
            }
            
        }
        return -1
    }
    return bfs(maps)
    
}
    

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    row = len(maps)
    col = len(maps[0])
    answer = 0
    visited = [[False] * col for _ in range(row)]
    visited[0][0] = True
    dist = 1
    queue = deque([(0,0,dist)])
    
    
    def dfs(maps) :
        dr = [-1, 1, 0 ,0]
        dc = [0,0, -1,1]
        while queue  : 
            r,c,dist = queue.popleft()
            if(r== row-1 and c == col -1):
                return dist
            for i in range(4) :
                nr, nc = r + dr[i], c+dc[i]
                if(nr>=0 and nr<row and nc >=0 and nc<col and not visited[nr][nc] and maps[nr][nc]!=0):
                    visited[nr][nc] = True
                    queue.append((nr,nc,dist+1))
        return -1
    
    
    return dfs(maps)
    
    